### PR TITLE
Document negative-input NaN behavior of PoissonNLLLoss

### LINKED ISSUE #184030: torch.nn.PoissonNLLLoss returns NaN for negative input when log_input=False
### 🐛 Describe the bug

`torch.nn.PoissonNLLLoss(log_input=False)` silently returns `NaN` for any input element that is negative. With `log_input=False` the loss computes `input - target * log(input + eps)`; the docstring states that input parameterises a Poisson rate (so `input >= 0`), but the constructor and forward path do not validate this, and `log` of a negative number propagates as `NaN` through the reduction.

The same hazard does not exist for `log_input=True`, where the loss is `exp(input) - target * input` (finite for any real `input`).

```python
import torch

torch.manual_seed(0)
loss = torch.nn.PoissonNLLLoss(log_input=False, reduction="none")
inp = torch.randn(2, 8)   # contains negatives
target = torch.rand(2, 8)
out = loss(inp, target)
print(out)
print("isnan:", torch.isnan(out).any().item())
```

Observed:

```
tensor([[   nan,    nan,    nan,    nan, 0.9605, 1.0289,    nan,    nan],
        [0.7972,    nan, 1.3502, 0.3507, 0.5128, 1.1580, 1.0831,    nan]])
isnan: True
```

Expected:

Either a `ValueError` rejecting negative input when `log_input=False`, or a documented note that callers must clamp / softplus the input first. The current behaviour (silent `NaN`) violates the Poisson-rate invariant implicit in the docstring formula `target ~ Poisson(input)`.

### Versions

```
Collecting environment information...
PyTorch version: 2.12.0a0+git71a2161
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.3 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: 18.1.3 (1ubuntu1)
CMake version: version 3.28.3
Libc version: glibc-2.39

Python version: 3.11.15 | packaged by conda-forge | (main, Mar  5 2026, 16:45:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.8.0-111-generic-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: 
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1660 SUPER
Nvidia driver version: 535.288.01
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           46 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  20
On-line CPU(s) list:                     0-19
Vendor ID:                               GenuineIntel
Model name:                              12th Gen Intel(R) Core(TM) i7-12700
CPU family:                              6
Model:                                   151
Thread(s) per core:                      2
Core(s) per socket:                      12
Socket(s):                               1
Stepping:                                2
CPU(s) scaling MHz:                      51%
CPU max MHz:                             4900.0000
CPU min MHz:                             800.0000
BogoMIPS:                                4224.00
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities ibpb_exit_to_user
Virtualization:                          VT-x
L1d cache:                               512 KiB (12 instances)
L1i cache:                               512 KiB (12 instances)
L2 cache:                                12 MiB (9 instances)
L3 cache:                                25 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-19
Vulnerability Gather data sampling:      Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Reg file data sampling:    Mitigation; Clear Register File
Vulnerability Retbleed:                  Not affected
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Mitigation; IBPB before exit to userspace

Versions of relevant libraries:
[pip3] numpy==2.4.4
[pip3] optree==0.19.0
[pip3] torch==2.12.0a0+git71a2161
[pip3] triton==3.6.0
[conda] numpy                       2.4.4            pypi_0              pypi
```

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

### PR DESCRIPTION
**Pull request by dparikh79**
Fixes #184030.

`PoissonNLLLoss` and the underlying `F.poisson_nll_loss` compute `input - target * log(input + eps)` when `log_input=False`. Any element of `input` smaller than `-eps` drives the `log(...)` to `NaN`, and the `NaN` propagates through the reduction. The docstring formula already shows `log(input + eps)`, but neither the `Module` docstring nor the functional docstring mentions the non-negativity requirement, so users feeding real-valued pre-activations (e.g. the output of a `Linear` with no positivity-enforcing activation) into the loss silently get `NaN` output, which is hard to root-cause from the visible API.

The reporter (#184030) suggested either rejecting negative input as a `ValueError` or documenting the contract. This PR takes the documentation path: behavior is unchanged, and the docstring update makes the input domain explicit so callers can clamp or `softplus` upstream as needed.

## What changed

- `torch/nn/functional.py::poisson_nll_loss`: extended the `input` argument description to call out the non-negativity requirement when `log_input=False`, the resulting `NaN` propagation, and the two standard workarounds (`torch.clamp(min=0)` or `torch.nn.functional.softplus`).
- `torch/nn/modules/loss.py::PoissonNLLLoss`: added a top-level `.. note::` block with the same contract, and explicitly noted that `log_input=True` is finite for any real input.

No code change; no behavior change; no new test (docs-only).

## How to verify

The reporter's repro still produces `NaN` (because behavior is unchanged), but the updated docstring now points the reader directly at the cause and the recommended fixes:

```python
import torch
torch.manual_seed(0)
loss = torch.nn.PoissonNLLLoss(log_input=False, reduction="none")
inp = torch.randn(2, 8)   # contains negatives
target = torch.rand(2, 8)
print(loss(inp, target))  # NaN-laced (as documented now)
print(loss(inp.clamp(min=0), target))  # finite (the documented fix)
```

Rendered Sphinx output for both files renders cleanly (`.. note::` and the extended `Args:` entry both follow the existing conventions in `loss.py`'s siblings).

## Notes for reviewers

Per `AGENTS.md` this PR was authored with AI assistance (Claude Code). I read the issue, traced `torch.poisson_nll_loss` from Python down to the C++ op, confirmed the formula matches `aten/src/ATen/native/Loss.cpp`, and chose the docs-only path so the change is risk-free vs. a runtime `ValueError` that would change the API contract. Happy to follow up with a stricter validation PR if maintainers prefer; this one is meant to land the cheaper half of the fix.

### CODE REVIEW

**Comment by albanD** [comment]
Thank you very much for sending a PR.
To avoid you spending time preparing a PR for a feature we're not sure we want to add, our usual process is to discuss on the issue before doing the PR.
In particular, make sure that the issue is marked "actionable" before sending a PR!
See https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions for details!
